### ISSUE TITLE
[AppBarLayout]when isIndexOutOfBoundsException

### ISSUE THREAD

**Issue by androidfei** [material-components/material-components-android#1390]
log:

java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
	at java.util.ArrayList.get(ArrayList.java:437)
	at com.google.android.material.appbar.AppBarLayout.a(AppBarLayout.java:12)
	at com.google.android.material.appbar.AppBarLayout$BaseBehavior.a(AppBarLayout.java:91)
	at com.google.android.material.appbar.AppBarLayout$BaseBehavior.b(AppBarLayout.java:2)
	at com.google.android.material.appbar.HeaderBehavior.b(HeaderBehavior.java:1)
	at com.google.android.material.appbar.AppBarLayout$BaseBehavior$a.onAnimationUpdate(AppBarLayout.java:3)
	at android.animation.ValueAnimator.animateValue(ValueAnimator.java:1547)
	at android.animation.ValueAnimator.animateBasedOnTime(ValueAnimator.java:1339)
	at android.animation.ValueAnimator.doAnimationFrame(ValueAnimator.java:1471)
	at android.animation.AnimationHandler.doAnimationFrame(AnimationHandler.java:146)
	at android.animation.AnimationHandler.access$100(AnimationHandler.java:37)
	at android.animation.AnimationHandler$1.doFrame(AnimationHandler.java:54)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:947)
	at android.view.Choreographer.doCallbacks(Choreographer.java:761)
	at android.view.Choreographer.doFrame(Choreographer.java:693)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:935)
	at android.os.Handler.handleCallback(Handler.java:873)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:214)
	at android.app.ActivityThread.main(ActivityThread.java:7073)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:964)

**Comment by ymarian** [comment]
Can you update the description following the template? A sample app that reproes the issue is always appreciated.

**Comment by drchen** [comment]
I'll close this issue for now due to no further info available. Please feel free to reopen it if this is still happening.

**Comment by kelthuzad13** [comment]
meet IndexOutOfBoundsException in onOffsetChanged

> 2025-05-14 03:13:38.761867  3918  3918 I BusinessFragment | onOffsetChanged

> 2025-05-14 03:13:38.762357  3918  3918 D AndroidRuntime: Shutting down VM
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: FATAL EXCEPTION: main
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: Process: xxxxxx, PID: 3918
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at java.util.ArrayList.get(ArrayList.java:437)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at com.google.android.material.appbar.AppBarLayout.onOffsetChanged(AppBarLayout.java:924)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at com.google.android.material.appbar.AppBarLayout$BaseBehavior.setHeaderTopBottomOffset(AppBarLayout.java:2100)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at com.google.android.material.appbar.AppBarLayout$BaseBehavior.setHeaderTopBottomOffset(AppBarLayout.java:1491)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at com.google.android.material.appbar.HeaderBehavior.setHeaderTopBottomOffset(HeaderBehavior.java:158)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at com.google.android.material.appbar.AppBarLayout$BaseBehavior.onLayoutChild(AppBarLayout.java:1844)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at com.google.android.material.appbar.AppBarLayout$Behavior.onLayoutChild(AppBarLayout.java:1472)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at com.google.android.material.appbar.AppBarLayout$BaseBehavior.onLayoutChild(AppBarLayout.java:1491)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at androidx.coordinatorlayout.widget.CoordinatorLayout.onLayout(CoordinatorLayout.java:918)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.View.layout(View.java:23894)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6420)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at androidx.constraintlayout.widget.ConstraintLayout.onLayout(ConstraintLayout.java:1873)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.View.layout(View.java:23894)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6420)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.View.layout(View.java:23894)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6420)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at androidx.constraintlayout.widget.ConstraintLayout.onLayout(ConstraintLayout.java:1873)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.View.layout(View.java:23894)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6420)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.View.layout(View.java:23894)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6420)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1891)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1729)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.widget.LinearLayout.onLayout(LinearLayout.java:1638)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.View.layout(View.java:23894)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6420)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at com.android.internal.widget.CustomFreeformDecorCaptionView.onLayout(CustomFreeformDecorCaptionView.java:354)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.View.layout(View.java:23894)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6420)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at com.android.internal.policy.DecorView.onLayout(DecorView.java:812)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.View.layout(View.java:23894)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.ViewGroup.layout(ViewGroup.java:6420)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:4205)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:3547)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2453)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:9472)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1231)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1239)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.Choreographer.doCallbacks(Choreographer.java:899)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.Choreographer.doFrame(Choreographer.java:832)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1214)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:942)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:201)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:288)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8002)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:551)
2025-05-14 03:13:38.777639  3918  3918 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:957)


in business code, we called removeOnOffsetChangedListener in OnOffsetChangedListener

```
private val offsetChangedListener = object : AppBarLayout.OnOffsetChangedListener {
            override fun onOffsetChanged(appBarLayout: AppBarLayout, verticalOffset: Int) {
                if (verticalOffset == 0) {

                } else if (abs(verticalOffset) == appBarLayout.totalScrollRange) {

                    Log.i(sometag, "onOffsetChanged")

                    appBarLayout.removeOnOffsetChangedListener(this)

                    // do business

                }
            }
        }
```


and  this is the code in AppbarLayout#onOffsetChanged

``` 
void onOffsetChanged(int offset) {
    currentOffset = offset;

    if (!willNotDraw()) {
      ViewCompat.postInvalidateOnAnimation(this);
    }

    // Iterate backwards through the list so that most recently added listeners
    // get the first chance to decide
    if (listeners != null) {
      for (int i = 0, z = listeners.size(); i < z; i++) {
        final BaseOnOffsetChangedListener listener = listeners.get(i);
        if (listener != null) {
          listener.onOffsetChanged(this, offset);
        }
      }
    }
  }
```

check this line 
> final BaseOnOffsetChangedListener listener = listeners.get(i);

if we removed OnOffsetChangedListener in the callback, the size in listeners will decreased by 1, and in the end of for loop, the **get** function will meet IndexOutOfBoundsException

in [this post](https://stackoverflow.com/questions/38056204/appbarlayout-how-to-removeonoffsetchangedlistener), there is a trick that we could move removeOnOffsetChangedListener into view's post, it can reduce the possibility to meet the crash. 

but it is still a hidden problem. hope it can be fixed in future release.
